src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3Connection.cs

```diff
@@ -0,0 +1,731 @@
+            Exception firstException = Interlocked.CompareExchange(ref _abortException, abortException, null);
+
+            if (firstException != null &&
+                NetEventSource.IsEnabled &&
+                !ReferenceEquals(firstException, abortException))
+            {
+                // Lost the race to set the field to another exception, so just trace this one.
+                Trace($"{nameof(abortException)}=={abortException}");
+                return;
+            }
+
+            // Stop sending requests to this connection.
+            _pool.InvalidateHttp3Connection(this);
+
+            Http3ErrorCode connectionResetErrorCode = (abortException as Http3ProtocolException)?.ErrorCode ?? Http3ErrorCode.ProtocolError;
+
+            lock (SyncObj)
+            {
+                // Set _lastProcessedStreamId != -1 to make ShuttingDown = true.
+                // It's possible GOAWAY is already being processed, in which case this would already be != -1.
+                if (_lastProcessedStreamId == -1)
+                {
+                    _lastProcessedStreamId = long.MaxValue;
+                }
+
+                // Abort the connection. This will cause all of our streams to abort on their next I/O.
+                if (_connection != null)
+                {
+                    _connection.Close((long)connectionResetErrorCode);
+                }
+
+                // Cancel any waiters.
+                CancelWaiters();
+
+                // Check here in case there are no active requests.
+                CheckForShutdown();
+            }
+        }
+
+        private void OnServerGoAway(long lastProcessedStreamId)
+        {
+            // Stop sending requests to this connection.
+            _pool.InvalidateHttp3Connection(this);
+
+            var streamsToGoAway = new List<Http3RequestStream>();
+
+            lock (SyncObj)
+            {
+                if (lastProcessedStreamId > _lastProcessedStreamId)
+                {
+                    // Server can send multiple GOAWAY frames.
+                    // Spec says a server MUST NOT increase the stream ID in subsequent GOAWAYs,
+                    // but doesn't specify what client should do if that is violated. Ignore for now.
+                    if (NetEventSource.IsEnabled)
+                    {
+                        Trace("HTTP/3 server sent GOAWAY with increasing stream ID. Retried requests may have been double-processed by server.");
+                    }
+                    return;
+                }
+
+                _lastProcessedStreamId = lastProcessedStreamId;
+
+                foreach (KeyValuePair<long, Http3RequestStream> request in _activeRequests)
+                {
+                    if (request.Key > lastProcessedStreamId)
+                    {
+                        streamsToGoAway.Add(request.Value);
+                    }
+                }
+
+                // Cancel any waiters.
+                CancelWaiters();
```

Comment:
Thoughts on sending GoAway on each stream before cancelling waiters?

Reply:
What's the goal?

Reply:
Nvrm, I misunderstood what these were waiting on. Thought these were active requests rather than pending requests.